pkg/common/cns-lib/volume/manager.go

```diff
@@ -184,161 +192,235 @@ func (m *defaultManager) ResetManager(ctx context.Context, vcenter *cnsvsphere.V
+				// Don't return if CreateVolume details can't be stored.
+				log.Warnf("failed to store CreateVolume details with error: %v", err)
 			}
 		}
-		// Get the taskInfo.
-		taskInfo, err = cns.GetTaskInfo(ctx, task)
-		if err != nil || taskInfo == nil {
-			log.Errorf("failed to get taskInfo for CreateVolume task from vCenter %q with err: %v",
-				m.virtualCenter.Config.Host, err)
-			return nil, err
+	}
+
+	// Get the taskInfo.
+	taskInfo, err := cns.GetTaskInfo(ctx, task)
+	if err != nil || taskInfo == nil {
+		log.Errorf("failed to get taskInfo for CreateVolume task with err: %v", err)
+		return nil, err
+	}
+
+	log.Infof("CreateVolume: VolumeName: %q, opId: %q", volNameFromInputSpec, taskInfo.ActivationId)
+	taskResult, err := getTaskResultFromTaskInfo(ctx, taskInfo)
+	if err != nil {
+		log.Errorf("failed to get task result for task %s and volume name %s with error: %v",
+			task.Reference().Value, volNameFromInputSpec, err)
+		return nil, err
+	}
+	volumeOperationRes := taskResult.GetCnsVolumeOperationResult()
+	if volumeOperationRes.Fault != nil {
+		// Validate if the volume is already registered.
+		resp, err := validateCreateVolumeResponseFault(ctx, volNameFromInputSpec, volumeOperationRes)
+		if err != nil {
+			volumeOperationDetails = createRequestDetails(volNameFromInputSpec, "", "", 0,
+				volumeOperationDetails.OperationDetails.TaskInvocationTimestamp, task.Reference().Value,
+				taskInfo.ActivationId, taskInvocationStatusError, err.Error())
+		} else {
+			volumeOperationDetails = createRequestDetails(volNameFromInputSpec, resp.VolumeID.Id, "", 0,
+				volumeOperationDetails.OperationDetails.TaskInvocationTimestamp, task.Reference().Value,
+				taskInfo.ActivationId, taskInvocationStatusSuccess, "")
 		}
-		log.Infof("CreateVolume: VolumeName: %q, opId: %q", volNameFromInputSpec, taskInfo.ActivationId)
-		// Get the taskResult.
-		taskResult, err := cns.GetTaskResult(ctx, taskInfo)
+		return resp, err
+	}
+	// Extract the CnsVolumeInfo from the taskResult.
+	resp, err := getCnsVolumeInfoFromTaskResult(ctx, m.virtualCenter, volNameFromInputSpec,
+		volumeOperationRes.VolumeId, taskResult)
+	if err != nil {
+		volumeOperationDetails = createRequestDetails(volNameFromInputSpec, "", "", 0,
+			volumeOperationDetails.OperationDetails.TaskInvocationTimestamp, task.Reference().Value,
+			taskInfo.ActivationId, taskInvocationStatusError, err.Error())
+	} else {
+		volumeOperationDetails = createRequestDetails(volNameFromInputSpec, resp.VolumeID.Id, "", 0,
+			volumeOperationDetails.OperationDetails.TaskInvocationTimestamp, task.Reference().Value,
+			taskInfo.ActivationId, taskInvocationStatusSuccess, "")
+	}
+	return resp, err
+
+}
 
+// createVolumeWithoutIdempotency invokes CNS CreateVolume without any idempotency handling. It stores task information in an in-memory map.
+func (m *defaultManager) createVolumeWithoutIdempotency(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (*CnsVolumeInfo, error) {
```

Comment:
Can you confirm that you are not making any logical change in this method?

Reply:
Yes there is no logical change. Only refactoring so that the functions can be reused in `createVolumeWithIdempotency`